### DISCUSSION TITLE
How to obtain control on a company

### DISCUSSION THREAD

**Discussion by DamsolDeveloper** [fattureincloud/api#474] [Q&A] [answered]
I'm a developer and I'm developing an app for another company.
I set my account of fatturincloud to start develop but then my app should be able to access data of the other company.
So i created the oauth schema and asked the holder of the company to go there and give access to my app.
I thought that this would have been enough to find the company as a "controlled_company" when I issue the userListCompanies APçI method but I dont see anything else than my company...
So the question is: how can I have another company inserted in my "controlled_companies" of in a list of companies that i can access in order ro handle theri data?
thanks

**Comment by valmoz** [comment, marked as answer]
Hello @DamsolDeveloper 
for clarity, let's call your company "A" and the other company "B".
If I'm not wrong, you created an app using company "A", and then you performed the OAuth flow with the created app on company "A" and "B", correct?

If that's the case, the OAuth step should have returned you two different tokens:

- The first one (let's call it "Z") is the one you obtained by performing the OAuth flow for company "A"
- The second one (let's call it "W") is the one that the holder of company "B" obtained by performing the OAuth flow for company "B".

In this situation, each token will be able to access the companies associated with the token itself:
- By performing API calls using the token "Z", you will be able to access company "A"
- By performing API calls using the token "W", you will be able to access company "B"

So if you need to perform operations on company "B" you must be able to collect the token "W" in some way. You cannot use the token "Z", even if the same app generated it, because that specific token is connected only to A and not B.

You can check the companies associated with a single token using the [List User Companies](https://developers.fattureincloud.it/api-reference#get-/user/companies) method.

I hope my explanation is clear. Let me know if you have further questions.

Thanks
Mauro

**Reply by DamsolDeveloper** [reply]
thanks (grazie, vedo che sei italiano se ok continuiamo in italiano)
Lo scenario che mi descritto è chiarissimo grazie.
Ti espongo la mia necessità: io ho bisogno di fare dei test con i dati reali della company B e non posso ogni volta che devo fare test chiedere al cliente (il capo di B) di fare l'auth... 
Mi chiedevo quindi se ci fosse un modo per avere un accesso permanente alla company B  (quelle che forse sono indicate come controlled_company) e pensavo che il solo far l'oauth mettesse la company B nella lista delle company gestibili da A. Ho capito male?
Esiste un modo per avere quindi un accesso permanente?
grazie mille

**Reply by valmoz** [reply]
@DamsolDeveloper 
deve semplicemente farsi passare access token e refresh token ottenuti dalla company B.
Che sia copiando e incollando i valori oppure utilizzando un meccanismo più automatizzato, ottenendo i token avrà accesso all'azienda desiderata.
Per chiarezza, l'access token viene utilizzato per effettuare le chiamate API e ha scadenza limitata, mentre il refresh token serve ad ottenere nuovi access token senza rieffettuare il flusso oauth e ha scadenza di un anno dall'ultimo utilizzo.

Mauro

**Reply by DamsolDeveloper** [reply]
grazie tutto chiaro.

**Comment by DamsolDeveloper** [comment]
chiudo
